File: practice/编译原理实验一/lexer/lexer/lexer.go
package lexer

import (
	"fmt"
	"unicode"
)

type Lexer struct {
	input       string
	pos         int
	line        int
	column      int
	currentChar rune
}

// 新增反向映射表（放在文件顶部）
var reverseKeywords = map[string]TokenType{
	"public":   PUBLIC,
	"class":    CLASS,
	"static":   STATIC,
	"void":     VOID,
	"main":     MAIN,
	"char":     CHAR,
	"int":      INT,
	"printf":   PRINTF,
	"scanf":    SCANF,
	"switch":   SWITCH,
	"case":     CASE,
	"default":  DEFAULT,
	"for":      FOR,
	"if":       IF,
	"else":     ELSE,
	"while":    WHILE,
	"do":       DO,
	"return":   RETURN,
	"break":    BREAK,
	"continue": CONTINUE,
}

func New(input string) *Lexer {
	l := &Lexer{
		input:  input,
		line:   1,
		column: 1,
	}
	if len(input) > 0 {
		l.currentChar = rune(input[0])
	}
	return l
}

func (l *Lexer) advance() {
	if l.currentChar == '\n' {
		l.line++
		l.column = 0
	}
	l.pos++
	if l.pos >= len(l.input) {
		l.currentChar = 0
	} else {
		l.currentChar = rune(l.input[l.pos])
	}
	l.column++
}

func (l *Lexer) skipWhitespace() {
	for unicode.IsSpace(l.currentChar) {
		l.advance()
	}
}

func (l *Lexer) readIdentifier() Token {
	startPos := l.pos
	startLine := l.line
	startCol := l.column

	for unicode.IsLetter(l.currentChar) || unicode.IsDigit(l.currentChar) {
		l.advance()
	}

	// 修正后的关键字识别逻辑
	value := l.input[startPos:l.pos]
	tokenType := IDENTIFIER
	if code, exists := reverseKeywords[value]; exists {
		tokenType = code
	}
	return Token{tokenType, value, startLine, startCol}
}

func (l *Lexer) readNumber() Token {
	startPos := l.pos
	startLine := l.line
	startCol := l.column

	for unicode.IsDigit(l.currentChar) {
		l.advance()
	}

	value := l.input[startPos:l.pos]
	return Token{NUMBER, value, startLine, startCol}
}

// 新增字符字面量解析方法
func (l *Lexer) readChar() Token {
	startLine := l.line
	startCol := l.column
	l.advance() // 跳过开始的单引号

	var value string
	if l.currentChar == '\\' { // 处理转义字符
		l.advance()
		switch l.currentChar {
		case 'n':
			value = "\n"
		case 't':
			value = "\t"
		case 'r':
			value = "\r"
		case '\'':
			value = "'"
		case '\\':
			value = "\\"
		default:
			panic(fmt.Sprintf("无效转义字符: \\%c 在行 %d:%d", l.currentChar, l.line, l.column))
		}
		l.advance()
	} else {
		value = string(l.currentChar)
		l.advance()
	}

	if l.currentChar != '\'' {
		panic(fmt.Sprintf("字符未闭合，起始于行 %d:%d", startLine, startCol))
	}
	l.advance() // 跳过闭合的单引号
	return Token{CharConst, value, startLine, startCol}
}
func (l *Lexer) readString() Token {
	startLine := l.line
	startCol := l.column
	l.advance() // Skip opening "

	var value string
	for l.currentChar != '"' && l.currentChar != 0 {
		if l.currentChar == '\\' {
			l.advance()
			switch l.currentChar {
			case 'n':
				value += "\n"
			case 't':
				value += "\t"
			case 'r':
				value += "\r"
			case '"':
				value += "\""
			case '\\':
				value += "\\"
			default:
				panic(fmt.Sprintf("Invalid escape: \\%c at %d:%d", l.currentChar, l.line, l.column))
			}
			l.advance()
		} else {
			value += string(l.currentChar)
			l.advance()
		}
	}

	if l.currentChar != '"' {
		panic(fmt.Sprintf("Unclosed string at %d:%d", startLine, startCol))
	}
	l.advance() // Skip closing "
	return Token{STRING, value, startLine, startCol}
}

func (l *Lexer) NextToken() Token {
	for l.currentChar != 0 {
		if unicode.IsSpace(l.currentChar) {
			l.skipWhitespace()
			continue
		}
		// 新增字符字面量处理
		if l.currentChar == '\'' {
			return l.readChar()
		}

		if l.currentChar == '"' {
			return l.readString()
		}

		if unicode.IsLetter(l.currentChar) {
			return l.readIdentifier()
		}

		if unicode.IsDigit(l.currentChar) {
			return l.readNumber()
		}

		// Handle operators and symbols
		currentChar := l.currentChar
		currentLine := l.line
		currentCol := l.column
		l.advance()

		switch currentChar {
		case '=':
			if l.currentChar == '=' {
				l.advance()
				return Token{EQUAL, "==", currentLine, currentCol}
			}
			return Token{ASSIGN, "=", currentLine, currentCol}
		case '!':
			if l.currentChar == '=' {
				l.advance()
				return Token{NOTEQ, "!=", currentLine, currentCol}
			}
			return Token{NOT, "!", currentLine, currentCol}
		case '<':
			if l.currentChar == '=' {
				l.advance()
				return Token{LessEqual, "<=", currentLine, currentCol}
			}
			return Token{LESS, "<", currentLine, currentCol}
		case '>':
			if l.currentChar == '=' {
				l.advance()
				return Token{GreatEqual, ">=", currentLine, currentCol}
			}
			return Token{GREAT, ">", currentLine, currentCol}
		case '+':
			return Token{PLUS, "+", currentLine, currentCol}
		case '-':
			return Token{MINUS, "-", currentLine, currentCol}
		case '*':
			return Token{MULTIPLY, "*", currentLine, currentCol}
		case '/':
			return Token{DIVIDE, "/", currentLine, currentCol}
		case ';':
			return Token{SEMICOLON, ";", currentLine, currentCol}
		case ',':
			return Token{COMMA, ",", currentLine, currentCol}
		case '(':
			return Token{LeftParen, "(", currentLine, currentCol}
		case ')':
			return Token{RightParen, ")", currentLine, currentCol}
		case '{':
			return Token{LeftBrace, "{", currentLine, currentCol}
		case '}':
			return Token{RightBrace, "}", currentLine, currentCol}
		case '[':
			return Token{LeftBrack, "[", currentLine, currentCol}
		case ']':
			return Token{RightBrack, "]", currentLine, currentCol}
		case ':':
			return Token{COLON, ":", currentLine, currentCol}
		default:
			panic(fmt.Sprintf("Unexpected character: %c at %d:%d", currentChar, currentLine, currentCol))
		}
	}

	return Token{EOF, "", l.line, l.column}
}

File: practice/PracticeLexer/src/main/java/cn/edu/ctgu/Lexer.java
package cn.edu.ctgu;
// Lexer.java

import java.util.HashMap;
import java.util.Map;

/**
 * 词法分析器核心类
 * 主要功能：
 * 1. 逐个字符读取输入字符串
 * 2. 识别不同词法单元模式
 * 3. 生成对应Token对象
 * 4. 跟踪源码位置信息
 */
public class Lexer {
    // 输入字符串和当前位置跟踪
    private final String input;   // 原始输入代码
    private int pos;              // 当前字符位置索引
    private int line;             // 当前行号（从1开始）
    private int column;           // 当前列号（从1开始）
    private char currentChar;     // 当前处理的字符


    // 关键字映射表（标识符到TokenType的映射）
    private static final Map<String, TokenType> KEYWORDS = new HashMap<>();

    static {
        // 基础关键字
        KEYWORDS.put("public", TokenType.PUBLIC);
        KEYWORDS.put("class", TokenType.CLASS);
        KEYWORDS.put("static", TokenType.STATIC);
        KEYWORDS.put("void", TokenType.VOID);
        KEYWORDS.put("int", TokenType.INT);

        // 新增关键字
        KEYWORDS.put("main", TokenType.MAIN);
        KEYWORDS.put("char", TokenType.CHAR);
        KEYWORDS.put("printf", TokenType.PRINTF);
        KEYWORDS.put("scanf", TokenType.SCANF);
        KEYWORDS.put("switch", TokenType.SWITCH);
        KEYWORDS.put("case", TokenType.CASE);
        KEYWORDS.put("default", TokenType.DEFAULT);
        KEYWORDS.put("for", TokenType.FOR);
        KEYWORDS.put("if", TokenType.IF);
        KEYWORDS.put("else", TokenType.ELSE);
        KEYWORDS.put("while", TokenType.WHILE);
        KEYWORDS.put("do", TokenType.DO);
        KEYWORDS.put("return", TokenType.RETURN);
        KEYWORDS.put("break", TokenType.BREAK);
        KEYWORDS.put("continue", TokenType.CONTINUE);
    }

    // 构造方法：初始化词法分析器
    public Lexer(String input) {
        this.input = input;
        this.pos = 0;                        // 从第一个字符开始
        this.line = 1;                        // 初始行号为1
        this.column = 1;                      // 初始列号为1
        // 初始化当前字符（处理空输入情况）
        this.currentChar = !input.isEmpty() ? input.charAt(0) : '\0';
    }
    private char peek() {
        if (pos + 1 >= input.length()) {
            return '\0';
        } else {
            return input.charAt(pos + 1);
        }
    }
    // 前进到下一个字符，更新位置信息
    private void advance() {
        // 处理换行符：行号增加，列号重置
        if (currentChar == '\n') {
            line++;
            column = 0;  // 下一轮advance会+1
        }

        pos++;           // 移动字符指针
        column++;        // 列号递增

        // 判断是否到达输入末尾
        if (pos >= input.length()) {
            currentChar = '\0';  // 结束符
        } else {
            currentChar = input.charAt(pos);  // 读取下一个字符
        }
    }

    // 跳过空白字符（包括空格、制表符、换行等）
    private void skipWhitespace() {
        // 循环处理连续空白字符
        while (Character.isWhitespace(currentChar)) {
            advance();
        }
    }

    // 读取标识符或关键字
    private Token readIdentifier() {
        // 记录起始位置信息
        int startPos = pos;
        int startLine = line;
        int startCol = column;

        // 持续读取字母和数字
        while (Character.isLetterOrDigit(currentChar)) {
            advance();
        }

        // 提取标识符字符串
        String value = input.substring(startPos, pos);
        // 判断是否是关键字
        TokenType type = KEYWORDS.getOrDefault(value, TokenType.IDENTIFIER);
        return new Token(type, value, startLine, startCol);
    }

    // 读取数字字面量
    private Token readNumber() {
        // 记录起始位置信息
        int startPos = pos;
        int startLine = line;
        int startCol = column;

        // 持续读取数字字符
        while (Character.isDigit(currentChar)) {
            advance();
        }

        // 提取数字字符串
        String value = input.substring(startPos, pos);
        return new Token(TokenType.NUMBER, value, startLine, startCol);
    }

    // 处理字符串字面量（带转义和闭合检查）
    private Token readString() {
        int startLine = line;
        int startCol = column;
        advance(); // 跳过开始的双引号

        StringBuilder sb = new StringBuilder();
        while (currentChar != '"' && currentChar != '\0') {
            if (currentChar == '\\') {
                advance(); // 处理转义字符
                switch (currentChar) {
                    case 'n' -> sb.append('\n');
                    case 't' -> sb.append('\t');
                    case 'r' -> sb.append('\r');
                    case '"' -> sb.append('"');
                    case '\\' -> sb.append('\\');
                    default -> throw new RuntimeException("无效转义字符: \\" + currentChar
                            + " 在行 " + line + ":" + column);
                }
                advance();
            } else {
                sb.append(currentChar);
                advance();
            }
        }

        if (currentChar != '"') {
            throw new RuntimeException("字符串未闭合，起始于行 " + startLine + ":" + startCol);
        }
        advance(); // 跳过闭合的双引号
        return new Token(TokenType.STRING, sb.toString(), startLine, startCol);
    }

    // 处理字符字面量（带转义和闭合检查）
    private Token readChar() {
        int startLine = line;
        int startCol = column;
        advance(); // 跳过开始的单引号

        char value;
        if (currentChar == '\\') { // 处理转义字符
            advance();
            value = switch (currentChar) {
                case 'n' -> '\n';
                case 't' -> '\t';
                case 'r' -> '\r';
                case '\'' -> '\'';
                case '\\' -> '\\';
                default -> throw new RuntimeException("无效转义字符: \\" + currentChar
                        + " 在行 " + line + ":" + column);
            };
            advance();
        } else {
            value = currentChar;
            advance();
        }

        if (currentChar != '\'') {
            throw new RuntimeException("字符未闭合，起始于行 " + startLine + ":" + startCol);
        }
        advance(); // 跳过闭合的单引号
        return new Token(TokenType.CHAR_CONST, String.valueOf(value), startLine, startCol);
    }

    // 主方法：生成下一个Token
    public Token nextToken() {
        // 循环处理直到生成有效Token或到达EOF
        while (currentChar != '\0') {
            // 跳过空白字符
            if (Character.isWhitespace(currentChar)) {
                skipWhitespace();
                continue;
            }

            // 处理字符串字面量
            if (currentChar == '"') {
                return readString();
            }

            // 处理字符字面量
            if (currentChar == '\'') {
                return readChar();
            }

            // 处理标识符（字母开头）
            if (Character.isLetter(currentChar)) {
                return readIdentifier();
            }

            // 处理数字字面量
            if (Character.isDigit(currentChar)) {
                return readNumber();
            }

            // 处理符号和运算符
            int currentLine = line;
            int currentCol = column;

            Token token = switch (currentChar) {
                case '=' -> {
                    if (peek() == '=') {
                        advance(); // 消耗当前'='
                        advance(); // 消耗下一个'='
                        yield new Token(TokenType.EQUAL, "==", currentLine, currentCol);
                    } else {
                        advance();
                        yield new Token(TokenType.ASSIGN, "=", currentLine, currentCol);
                    }
                }
                case '!' -> {
                    if (peek() == '=') {
                        advance();
                        advance();
                        yield new Token(TokenType.NOTEQ, "!=", currentLine, currentCol);
                    } else {
                        advance();
                        yield new Token(TokenType.NOT, "!", currentLine, currentCol);
                    }
                }
                case '<' -> {
                    if (peek() == '=') {
                        advance();
                        advance();
                        yield new Token(TokenType.LESS_EQUAL, "<=", currentLine, currentCol);
                    } else {
                        advance();
                        yield new Token(TokenType.LESS, "<", currentLine, currentCol);
                    }
                }
                case '>' -> {
                    if (peek() == '=') {
                        advance();
                        advance();
                        yield new Token(TokenType.GREAT_EQUAL, ">=", currentLine, currentCol);
                    } else {
                        advance();
                        yield new Token(TokenType.GREAT, ">", currentLine, currentCol);
                    }
                }
                case '+' -> new Token(TokenType.PLUS, "+", currentLine, currentCol);
                case '-' -> new Token(TokenType.MINUS, "-", currentLine, currentCol);
                case '*' -> new Token(TokenType.MULTIPLY, "*", currentLine, currentCol);
                case ':' -> new Token(TokenType.COLON, ":", currentLine, currentCol);
                case '/' -> new Token(TokenType.DIVIDE, "/", currentLine, currentCol);
                case ';' -> new Token(TokenType.SEMICOLON, ";", currentLine, currentCol);
                case ',' -> new Token(TokenType.COMMA, ",", currentLine, currentCol);
                case '(' -> new Token(TokenType.LEFT_PAREN, "(", currentLine, currentCol);
                case ')' -> new Token(TokenType.RIGHT_PAREN, ")", currentLine, currentCol);
                case '{' -> new Token(TokenType.LEFT_BRACE, "{", currentLine, currentCol);
                case '}' -> new Token(TokenType.RIGHT_BRACE, "}", currentLine, currentCol);
                case '[' -> new Token(TokenType.LEFT_BRACK, "[", currentLine, currentCol);
                case ']' -> new Token(TokenType.RIGHT_BRACK, "]", currentLine, currentCol);
                default -> throw new RuntimeException("非法字符: " + currentChar + " 位于行 " + currentLine + ":" + currentCol);
            };

            advance();
            return token;
        }


        // 输入结束返回EOF标记
        return new Token(TokenType.EOF, "", line, column);
    }
}